x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/index/IndexResolver.java

```diff
@@ -590,4 +646,134 @@ public void resolveAsSeparateMappings(String indexWildcard, String javaRegex, bo
         foundIndices.sort(Comparator.comparing(EsIndex::name));
         return foundIndices;
     }
+
+
+    /*
+     * Checks if the field is valid (same type and same capabilities - searchable/aggregatable) across indices belonging to a list
+     * of aliases.
+     * A field can look like the example below (generated by field_caps API).
+     *   "name": {
+     *       "text": {
+     *           "type": "text",
+     *           "searchable": false,
+     *           "aggregatable": false,
+     *           "indices": [
+     *               "bar",
+     *               "foo"
+     *           ],
+     *           "non_searchable_indices": [
+     *               "foo"
+     *           ]
+     *       },
+     *       "keyword": {
+     *           "type": "keyword",
+     *           "searchable": false,
+     *           "aggregatable": true,
+     *           "non_aggregatable_indices": [
+     *               "bar", "baz"
+     *           ]
+     *       }
+     *   }
+     */
+    private static Map<String, InvalidMappedField> getInvalidFieldsForAliases(String fieldName, Map<String, FieldCapabilities> types,
+            ImmutableOpenMap<String, List<AliasMetaData>> aliases) {
+        if (aliases == null || aliases.isEmpty()) {
+            return emptyMap();
+        }
+        Map<String, InvalidMappedField> invalidFields = new HashMap<>();
+        Map<String, Set<String>> typesErrors = new HashMap<>();
+        Map<String, Set<String>> aliasToIndices = new HashMap<>(); // map with aliases and their list of indices
+        
+        Iterator<ObjectObjectCursor<String, List<AliasMetaData>>> iter = aliases.iterator();
+        while (iter.hasNext()) {
+            ObjectObjectCursor<String, List<AliasMetaData>> index = iter.next();
+            for (AliasMetaData aliasMetaData : index.value) {
+                String aliasName = aliasMetaData.alias();
+                typesErrors.putIfAbsent(aliasName, new HashSet<>());
```

Comment:
this creates a HashSet regardless of the key presence - better to do the check

Reply:
Actually, this was intended like this.  See [this line](https://github.com/elastic/elasticsearch/pull/53525/files#diff-e133dec5642088d77645639c836c858aR715) where the field type for this field belonging to this alias is added to the list of unique field types.  And [here](https://github.com/elastic/elasticsearch/pull/53525/files#diff-e133dec5642088d77645639c836c858aR725) the size of the set is checked to see if there is more than one field type across all indices belonging to this alias. If there is more than one field type, then the field is considered "invalid" for this alias.

Reply:
I don't follow - `putIfAbsent` does `if (key == null) put(key, value)` - I'm pointing out that a new `HashSet` is created every time.
What I'm suggesting:
```
if (key == null) {
   put(key, new HashSet())
}
```

Reply:
I am aware of the `putIfAbsent` functionality. I _think_ I can simplify the logic a bit.